File: backend/types/world.go
package types

import (
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"strings"

	"gopkg.in/yaml.v3"
)

type World struct {
	WorldMap WorldMap
	Turn     int
	Users    []User
	Economy  Economy
	Society  Society
}

func CreateWorld() *World {
	return &World{
		Users:    Users,
		Turn:     0,
		WorldMap: WorldMapTemplate,
		Economy:  EconomyTemplate,
		Society:  SocietyTemplate,
	}
}

// Generic function to load YAML config into any struct type
func loadConfig[T any]() *T {
	fullTypeName := reflect.TypeOf((*T)(nil)).Elem().String()
	parts := strings.Split(fullTypeName, ".")        // Remove package name
	typeName := strings.ToLower(parts[len(parts)-1]) // Get only struct name
	configPath := filepath.Join("config", typeName+".yaml")

	data, err := os.ReadFile(configPath)
	if err != nil {
		fmt.Printf("failed to read config file: %s\n", err)
		return nil
	}

	var config T
	if err := yaml.Unmarshal(data, &config); err != nil {
		fmt.Printf("failed to parse config: %s\n", err)
		return nil
	}

	return &config
}

File: frontend/src/types/world.ts
import { Business, Economy } from "./economy";
import { Society } from "./society";
import { User } from "./user";

export interface World {
    WorldMap: WorldMap;
    Turn: number;
    Users: User[];
    Economy: Economy;
    Society: Society;
}

export interface WorldMap {
    Zones: Zone[];
    Territories: Territory[];
}

export interface Zone {
    ZoneId: string;
    Name: string;
    Type: string;
    X: number;
    Y: number;
    Width: number;
    Height: number;
    BusinessId: string;
}

export interface Territory {
    TerritoryId: string;
    Name: string;
    ControlledBy: string | null;
    Zones: string[];
}